### PR TITLE
Add background color support to Columns block

### PR DESCRIPTION
**Pull request by richtabor**
## Description
Add background color support to the core Columns block, in support of #16660. This uses the same method employed by the core Group block.

## How has this been tested?
Tested in supported browsers and ensured color classes and custom colors are applied to the columns block within the editor and the front-end. 

## Screenshots 
colors

## Types of changes
**New feature:** A new Color Settings panel for the Columns block, allowing folks to add a background color to their Column blocks.  

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation.

### CODE REVIEW

**Comment by CreativeDive** [comment]
@richtabor Nice work. But how we can change the text color in relation to the background color?

**Comment by richtabor** [comment]
> @richtabor Nice work. But how we can change the text color in relation to the background color?

It's tricky when we're styling nested blocks. I lean towards letting child blocks handle their own color styling, instead of having a priority style system. For example, you'd apply a background to the whole Columns block, but select your headings and color - then select your paragraphs and color.

**Review by mapk — APPROVED** [review]
Great work! 😍 I tested this and it worked just right. :shipit:  if the technical feedback is all good in the hood.

On a side note, I really want to change background colors on individual columns within the Columns block.

**Comment by ZebulanStanphill** [comment]
I think this could be refactored to use the new `useColors` hook.

Also, I agree that background color options for individual Column blocks would be a cool addition for another PR.

**Comment by richtabor** [comment]
@ZebulanStanphill I've updated the block to use `useColors` now. Mind taking a look to double-check? Thanks!

File: packages/block-library/src/columns/edit.js [hunk 1]
```diff
@@ -62,12 +63,30 @@ function ColumnsEditContainer( {
...
+	const {
+		BackgroundColor,
+		InspectorControlsColorPanel,
+	} = __experimentalUseColors(
+		[
+			{ name: 'backgroundColor', className: 'has-background' },
+		],
+		{
+			contrastCheckers: [ { backgroundColor: false } ],
+		}
```
Comment (ZebulanStanphill) [inline packages/block-library/src/columns/edit.js]: I don't understand what the purpose of including this is?

**Review by ZebulanStanphill — COMMENTED** [review]
One minor comment, but other than that, everything seems to be working fine.

**Comment by shaunandrews** [comment]
I’d vote to allow text color as well. Without it it’s a little confusing how it all works - I rarely trust magic. This same discussion is happening for the Group block: https://github.com/WordPress/gutenberg/pull/19181

**Comment by richtabor** [comment]
@shaunandrews I wrote a [comment](https://github.com/WordPress/gutenberg/pull/19181#issuecomment-567065045) on #19181 to highlight a concern with having a top-level `TextColor` control on a parent block (Columns, Group):

I'd appreciate your thoughts on it. :)

**Comment by richtabor** [comment]
> I’d vote to allow text color as well. Without it it’s a little confusing how it all works - I rarely trust magic. This same discussion is happening for the Group block: #19181

Also, what magic are you referring to? The auto contrast applied by the TwentyTwenty theme?

**Comment by shaunandrews** [comment]
> Also, what magic are you referring to? The auto contrast applied by the TwentyTwenty theme?

Yes, exactly. Anytime the editor tries to do something _for_ me, its attempting to be clever. Software is rarely great at being clever, and magic doesn't exist. I see this auto-text-color thing as an attempt at clever magic. For many use-cases, it might be a very clever solution that works — but for many other user-cases, it will just be confusing magic if it can't be overridden.

**Comment by richtabor** [comment]
> Yes, exactly. Anytime the editor tries to do something _for_ me, its attempting to be clever.

Well that's per the theme, not from the color palette control within Gutenberg. Most themes don't do that actually.

**Comment by BinaryMoon** [comment]
Currently you can change the background colour on columns, but this doesn't work properly when the column alignment is changed. Center vertical alignment and only the center section is coloured, not the whole column.

Can this be fixed here, or should it be a separate bug report?

**Comment by richtabor** [comment]
> Currently you can change the background colour on columns, but this doesn't work properly when the column alignment is changed. Center vertical alignment and only the center section is coloured, not the whole column.

Could you take a screen recording of this/provide browser details? I'm not getting the same behavior currently: 

ScreenFlow

**Comment by BinaryMoon** [comment]
@richtabor Hey Rich - I'm talking about individual columns, rather than the entire block. As I said, it may be (is likely) unrelated.

I made a little demo here: https://demo.prothemedesign.com/wordpress/gutenberg/2019/12/20/column-background-bug/

I expect the column background to stretch full height. Not to just wrap the content.

**Comment by ZebulanStanphill** [comment]
@BinaryMoon Yeah, I think that's a separate matter from this PR. This PR only concerns background color for the entire Columns block.

(Personally, I think it's debatable whether or not individual columns should be full-height.)

**Comment by BinaryMoon** [comment]
@ZebulanStanphill Thanks - I've added a new issue: https://github.com/WordPress/gutenberg/issues/19295

**Comment by mapk** [comment]
Seems like the questions have all been answered and clarified here. @richtabor if you're around, how about we get this merged?

**Comment by ZebulanStanphill** [comment]
I noticed an issue. If the Column blocks have nested Paragraph blocks and you use the custom color picker and drag the dot all the way to the top and try dragging beyond that, this happens:
image

This does not happen with the Group block.

**Comment by richtabor** [comment]
> I noticed an issue. If the Column blocks have nested Paragraph blocks and you use the custom color picker and drag the dot all the way to the top and try dragging beyond that, this happens:

Hey @ZebulanStanphill, are you still experiencing this? Here's a screencast I recorded trying to emulate that. If you are, could you record one too? Thanks!

ScreenFlow

**Comment by ZebulanStanphill** [comment]
@richtabor Yeah, there are no more block validation errors. The color picker pop-up does close when I finish dragging outside of its boundaries, though that's only a minor issue.

**Comment by richtabor** [comment]
Good deal @ZebulanStanphill! I'd say this is ready for merge then. 🎉

**Comment by karmatosed** [comment]
Thanks for all the work on this everyone. I am going to merge it and excited to see it in 🙌

**Pull request merged**